src/inmanta/server/services/orchestrationservice.py

```diff
@@ -63,277 +63,204 @@
+            env_id,
+            base_version,
+            version,
+            rids_in_partial_compile,
+            updated_resource_sets,
+            deleted_resource_sets,
+            updated_and_shared_resources_old,
+            rids_deleted_resource_sets,
+        )
 
-    def resource_changed_resource_set(self) -> bool:
+    def merge_updated_and_shared_resources(
+        self, updated_and_shared_resources: abc.Sequence[data.Resource]
+    ) -> Dict[ResourceIdStr, data.Resource]:
         """
-        return true if the resource_set in new_resource and old_resource are not the same
+        :param updated_and_shared_resources: The resources that are part of the partial compile.
+        :returns: The subset of resources in the new version of the configuration model that belong to the shared resource set
+                  or a resource set that is updated by this partial compile.
         """
-        assert self.old_resource is not None
-        return self.new_resource.resource_set != self.old_resource.resource_set
-
-
-@dataclasses.dataclass(frozen=True)
-class MergedModel:
-    """
-    A class containing the result of merging an old version of a configuration model with a partial model.
-    """
+        shared_resources = {r.resource_id: r for r in updated_and_shared_resources if r.resource_set is None}
+        updated_resources = {r.resource_id: r for r in updated_and_shared_resources if r.resource_set is not None}
+        shared_resources_merged = {r.resource_id: r for r in self._merge_shared_resources(shared_resources)}
+        result = {**updated_resources, **shared_resources_merged}
+        self._validate_constraints(list(result.values()))
+        return result
 
-    resources: list[dict[str, object]]
-    resource_states: dict[ResourceIdStr, Literal[ResourceState.available, ResourceState.undefined]]
-    resource_sets: dict[ResourceIdStr, Optional[str]]
-    unknowns: List[data.UnknownParameter]
+    def _validate_constraints(self, new_updated_and_shared_resources: abc.Sequence[data.Resource]) -> None:
+        """
+        Validate whether the new updated and shared resources that results from the merging the old version of the model
+        with resources of the partial compile, are compliant with the constraints of a partial compile.
 
+        :param new_updated_and_shared_resources: The resources that have to be validated.
+        """
+        for res in new_updated_and_shared_resources:
+            if res.resource_id not in self.updated_and_shared_resources_old:
+                continue
+            matching_resource_old_model = self.updated_and_shared_resources_old[res.resource_id]
 
-class PartialUpdateMerger:
-    """
-    This class is used to merge the result of a partial compile with previous resources and resource_sets. Takes a partial spec
-    that should be applied on a given base version to form the given new version.
+            if res.resource_set != matching_resource_old_model.resource_set:
```

Comment:
This works for movement between sets we have in memory, but what prevents movement between sets we don't have in memory? 

is there a database constraint for this? 

if so, why keep this duplicate? (I'm not implying the duplicate must go, I just want to understand why it exists)

Reply:
The resource table has a primary key on `(environment, model, resource_id)`. So we will get a primary key violation when there is movement between resource sets that are not in memory.